File: backend/app/models/cv_schema.py
from pydantic import BaseModel, Field


class ProcessingMetadata(BaseModel):
    strategy: str = "heuristic"
    provider: str = ""
    model: str = ""
    warnings: list[str] = Field(default_factory=list)


class CVBasics(BaseModel):
    full_name: str = ""
    headline: str = ""
    email: str = ""
    phone: str = ""
    location: str = ""
    linkedin: str = ""
    website: str = ""


class CVSkill(BaseModel):
    name: str
    category: str = "general"
    proficiency: str = "unknown"


class CVExperience(BaseModel):
    company: str = ""
    title: str = ""
    location: str = ""
    start_date: str = ""
    end_date: str = ""
    duration_months: int = 0
    bullets: list[str] = Field(default_factory=list)
    achievements: list[str] = Field(default_factory=list)
    technologies: list[str] = Field(default_factory=list)


class CVEducation(BaseModel):
    institution: str = ""
    degree: str = ""
    field_of_study: str = ""
    start_date: str = ""
    end_date: str = ""
    details: list[str] = Field(default_factory=list)


class CVCertification(BaseModel):
    name: str
    issuer: str = ""
    date: str = ""


class CVLanguage(BaseModel):
    name: str
    level: str = ""


class CVProject(BaseModel):
    name: str
    description: str = ""
    technologies: list[str] = Field(default_factory=list)
    url: str = ""


class StructuredCV(BaseModel):
    basics: CVBasics = Field(default_factory=CVBasics)
    summary: str = ""
    skills: list[CVSkill] = Field(default_factory=list)
    experience: list[CVExperience] = Field(default_factory=list)
    education: list[CVEducation] = Field(default_factory=list)
    certifications: list[CVCertification] = Field(default_factory=list)
    languages: list[CVLanguage] = Field(default_factory=list)
    projects: list[CVProject] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)
    raw_text: str = ""
    meta: ProcessingMetadata = Field(default_factory=ProcessingMetadata)

File: frontend/lib/cv-schema.ts
import { z } from "zod";

export const ProcessingMetadataSchema = z.object({
  strategy: z.string(),
  provider: z.string(),
  model: z.string(),
  warnings: z.array(z.string()),
});

export const StructuredCVSchema = z.object({
  basics: z.object({
    full_name: z.string(),
    headline: z.string(),
    email: z.string(),
    phone: z.string(),
    location: z.string(),
    linkedin: z.string(),
    website: z.string(),
  }),
  summary: z.string(),
  skills: z.array(
    z.object({ name: z.string(), category: z.string(), proficiency: z.string() }),
  ),
  experience: z.array(
    z.object({
      company: z.string(),
      title: z.string(),
      location: z.string(),
      start_date: z.string(),
      end_date: z.string(),
      duration_months: z.number(),
      bullets: z.array(z.string()),
      achievements: z.array(z.string()),
      technologies: z.array(z.string()),
    }),
  ),
  education: z.array(
    z.object({
      institution: z.string(),
      degree: z.string(),
      field_of_study: z.string(),
      start_date: z.string(),
      end_date: z.string(),
      details: z.array(z.string()),
    }),
  ),
  certifications: z.array(
    z.object({ name: z.string(), issuer: z.string(), date: z.string() }),
  ),
  languages: z.array(z.object({ name: z.string(), level: z.string() })),
  projects: z.array(
    z.object({
      name: z.string(),
      description: z.string(),
      technologies: z.array(z.string()),
      url: z.string(),
    }),
  ),
  keywords: z.array(z.string()),
  raw_text: z.string(),
  meta: ProcessingMetadataSchema,
});
